### PR TITLE
Do a better job of minimizing building

### LINKED ISSUE #167: Incremental builds re-do unnecessary work
This is the output of make after successful make:

```
make[1]: Nothing to be done for `MKDIR'.
Updating generated files for nifake
Creating Wheel distribution for nifake
Creating Source distribution for nifake
make[1]: Nothing to be done for `MKDIR'.
Updating generated files for nidmm
Creating Wheel distribution for nidmm
Creating Source distribution for nidmm
make[1]: Nothing to be done for `MKDIR'.
Updating generated files for nimodinst
Creating Wheel distribution for nimodinst
Creating Source distribution for nimodinst
Kirsch0:nimi-python kirsch$ 
```

These steps are unecessary. We need to add the right dependencies to prevent the unnecessary work. This will be more annoying as we expand nimi-python to contain more drivers.

For reference, we fixed this for unit tests:
* https://github.com/ni/nimi-python/issues/136
* https://github.com/ni/nimi-python/pull/165

### PR DESCRIPTION
**Pull request by texasaggie97**
[X] This contribution adheres to [CONTRIBUTING.md]
### What does this Pull Request accomplish?
* Fixes #167 
* Create file in LOG_DIR to track when a build step has completed successfully
  * Create variables in defines.mak for all the tracking files
* All targets except test & html now use this mechanism
  * sphinx already does a good job of not rebuilding when not required
  * `make test` should always run tests

### Why should this Pull Request be merged?
* Developer effenciency

### What testing has been done?
* Travis

### CODE REVIEW

File: build/defines.mak [hunk 1]
```diff
@@ -69,4 +69,10 @@ DEFAULT_RST_FILES_TO_GENERATE := \
                      attributes.rst \
                      functions.rst \
 
+# Files for tracking parts of the build
```
Comment (marcoskirsch) [inline build/defines.mak]: I think the names of these variables / files should make it obvious that they are related and that they are used for the same purpose.

How about _DONE / done ?

Comment (marcoskirsch) [reply build/defines.mak]: GENERATED_FILES_COPIED would become GENERATED_FILES_COPY_DONE

File: build/rules.mak [hunk 2]
```diff
@@ -60,14 +60,14 @@ module: $(MODULE_FILES)
 
 $(UNIT_TEST_FILES): $(MODULE_FILES) $(RST_FILES)
 
-unit_tests: $(UNIT_TEST_FILES)
+unit_tests: $(UNIT_TESTS_PASSED)
 
-$(LOG_DIR)/tests_passed: $(UNIT_TEST_FILES)
+$(UNIT_TESTS_PASSED): $(UNIT_TEST_FILES)
 	$(call trace_to_console, "Running pytest",$@)
```
Comment (marcoskirsch) [inline build/rules.mak]: I did realize that in the output of make, this prints:

` Running pytest: bin/nifake/log/tests_passed
`

It's a little bit awkward...

File: build/rules.mak [hunk 3]
```diff
@@ -77,13 +77,23 @@ $(OUTPUT_DIR)/setup.py: $(TEMPLATE_DIR)/setup.py.mako
...
+	$(_hide_cmds)$(call log_command,rm $@)
 	$(_hide_cmds)$(call log_command,cd $(OUTPUT_DIR) && python3 setup.py sdist $(LOG_OUTPUT) $(LOG_DIR)/sdist.log)
+	$(_hide_cmds)$(call log_command,touch $@)
+
+wheel: $(WHEEL_BUILT)
 
-wheel: $(OUTPUT_DIR)/setup.py $(OUTPUT_DIR)/README.rst $(MODULE_FILES) $(LOG_DIR)/tests_passed
+$(WHEEL_BUILT): $(OUTPUT_DIR)/setup.py $(OUTPUT_DIR)/README.rst $(MODULE_FILES) $(UNIT_TESTS_PASSED)
 	$(call trace_to_console, "Creating wheel",$(OUTPUT_DIR)/dist)
+	$(_hide_cmds)$(call log_command,touch $@)
```
Comment (marcoskirsch) [inline build/rules.mak]: Lot's of duplication for this touch - rm - do something - touch pattern.
We should put in a function.

File: generated/nidmm/setup.py [hunk 4]
```diff
@@ -28,7 +28,7 @@ def read_contents(file_to_read):
 setup(
     name=pypi_name,
     zip_safe=False,
-    version='0.1.0.dev3',
+    version='0.1.0.dev4',
```
Comment (marcoskirsch) [inline generated/nidmm/setup.py]: Was this a manual change?
Seems unrelated to PR.

File: build/defines.mak (hunk 1 above)
Comment (texasaggie97) [reply build/defines.mak]: _DONE

File: build/rules.mak (hunk 2 above)
Comment (texasaggie97) [reply build/rules.mak]: We can change this to something better in a later change, after deciding what "better" would mean for this.

File: build/rules.mak (hunk 3 above)
Comment (texasaggie97) [reply build/rules.mak]: Moved into make_with_tracking_file. I am open to a better name.

File: generated/nidmm/setup.py (hunk 4 above)
Comment (texasaggie97) [reply generated/nidmm/setup.py]: I build .dev3 and pushed to PyPI. This updates the config['version'] but that hadn't been built into the generated files yet.

File: build/defines.mak (hunk 1 above)
Comment (marcoskirsch) [reply build/defines.mak]: we should rename the file name as well, right?

File: build/rules.mak [hunk 5]
```diff
@@ -29,6 +29,13 @@ define trace_to_console
 	@echo "$(shell printf '%15s' $1): $(subst $(CURRENT_DIR)/,,$2)"
 endef
 
+define make_with_tracking_file
```
Comment (marcoskirsch) [inline build/rules.mak]: can you add documentation?

File: build/rules.mak [hunk 6]
```diff
@@ -104,16 +112,18 @@ test: $(TOX_INI)
 	$(call trace_to_console, "Running tox",$(OUTPUT_DIR))
 	$(_hide_cmds)$(call log_command,cd $(OUTPUT_DIR) && set DRIVER=$(DRIVER) && tox)
 
-flake8: $(TOX_INI)
-	$(call trace_to_console, "Running flake",$(OUTPUT_DIR))
-	$(_hide_cmds)$(call log_command,cd $(OUTPUT_DIR) && tox -e flake8)
+update_generated_files: $(GENERATED_FILES_DONE)
 
-update_generated_files: $(MODULE_FILES) $(OUTPUT_DIR)/setup.py
+# Can't use make_with_tracking_file since there are multiple commands
```
Comment (marcoskirsch) [inline build/rules.mak]: Maybe we should address this by breaking this up :/

File: build/defines.mak (hunk 1 above)
Comment (texasaggie97) [reply build/defines.mak]: I am working on another change to do most of this through tox directly with a single call, so these will be going away.
